Doors platforms triggers

victorbstan opened the pull request:
### Why?
Doors, platforms and triggers needed some updates to bring their feature set closer to parity with Quake game features

### What?
* Doors auto-close after being triggered (TODO: don't close if something is blocking the door)
* Doors & Platforms auto generate trigger volumes if they are not targets or require keys
* Platforms auto generate triggers if they are targets, after they have been triggered once
* TODO: platforms need to return to their initial position unless `wait -1` or `toggle` on. (currently, they are, but the target position is closed, when it should be open, so they return to the wrong position)

victorbstan committed:
Update .gitignore

victorbstan committed:
WIP door trigger area autogenerate

victorbstan committed:
doors setup trigger volumes after entities initialized

victorbstan committed:
DOORS, check if key is required, emit messages for key doors

victorbstan committed:
DOORS: trigers generated for platforms without targetnames

victorbstan committed:
DOOR platform generates autotarget, door auto-closes,

victorbstan commented on quake1_example/class/_func.gd:
```diff
@@ -7,8 +7,16 @@ func _get_properties(dict : Dictionary) : props = dict
 			aabb = aabb.merge(m.get_aabb())
 	aabb.position += global_position
+
+func _can_create_trigger() -> bool : return false;
```
There might be a better way than using this flag; using different class inheritance, for now, this does the trick IMHO

victorbstan replied:
This flag is mostly to get around secret doors that should not generate auto-triggers, but maybe a symptom of putting too much functionality into shared methods

victorbstan commented on quake1_example/class/func_door.gd:
```diff
@@ -115,27 +205,27 @@ func _get_sounds(sounds : int) :
 func _get_sound_index_loop() -> int : return 0
 func _get_sound_index_motion_end() -> int : return 1
-
-func _motion_f(destroy_tween : bool = false) :
```
renamed to _motion_end

victorbstan committed:
revert unnecessary change

victorbstan committed:
fix secret door triggering bug, fix door message display bug

victorbstan commented:
Note, platforms should probably not inherit from doors

victorbstan commented:
This has a lot of changes and the more I review it, the more I think I want to refactor it...

gongpha commented on quake1_example/class/func_door.gd:
```diff
@@ -152,44 +242,79 @@ func _audf() :
+
+
+func _requires_key() -> bool :
+	print('SF: ', props.get('spawnflags'))
```
Are you forget to remove this line ?

victorbstan replied:
Yep :/ will remove

victorbstan commented on quake1_example/class/_func.gd:
```diff
@@ -7,8 +7,16 @@ func _get_properties(dict : Dictionary) : props = dict
+func _can_create_trigger() -> bool : return false;
+func _create_trigger() -> void : pass
+
+func _touch_horizontal(nor : Vector3) -> bool :
```
Just some helper functions can be used to check if player touch sides or top of object

victorbstan commented on quake1_example/class/func_door.gd:
```diff
@@ -4,6 +4,9 @@ class_name QmapbspQuakeFunctionDoor
 # this is written from scratch. No QuakeC code was applied here
 # maybe not accurate to the original approach
 
+const GOLD_KEY_MESSAGE = 'You require the Gold Key!'
```
Different themes (tech, wizard, etc.) should have different descriptions, for example tech world uses red and blue key cards instead of gold and silver keys... should probably add those and add code that distinguishes between worlds at some point...

victorbstan commented on quake1_example/class/func_plat.gd:
```diff
@@ -1,27 +1,46 @@
+	
 func _player_touch(p : QmapbspQuakePlayer, pos : Vector3, nor : Vector3) :
-	if open and nor.y > 0.9 :
+	if not open and _touch_horizontal(nor) :
```
If the platform is "up"/open and player touches it's side, it will trigger it, so it returns back down

victorbstan commented on quake1_example/viewer.gd:
```diff
@@ -103,6 +109,7 @@ func play_by_mapname(mapname : String, no_console : bool = false) -> bool :
 	parser = QmapbspWorldImporterQuake1.new()
 	parser.viewer = self
 	map = QmapbspQuakeWorld.new()
+	map.name = &'map'
```
Not related, but through it would be nicer in the tree view than `@@123`...

gongpha commented:
I confirmed this PR fixes a lot of movable brushes (e.g. e1m1, e1m8) and makes doors open properly from the map logic.

Some problematic brushes (most of them are platform brushes, e.g. e4m2, that one platform brush under the water pool) are still broken, though. (According to #15)

victorbstan committed:
removed gotdot git plugin

victorbstan committed:
Added spawnflags consts, check const by name, refactored doors to separate logic, updated platform and secret doors

victorbstan committed:
Merge branch 'master' into doors-platforms-triggers

victorbstan commented:
* Platforms should now be placed correctly and extend and retract after being triggered (still some things to work on)
* Separated doors, secret, plats, from sharing code (no inheritance) in order to prevent bugs and simplify development for each
* Secret doors should now open and close properly in any direction
* Added consts for spawnflags to funcs and mechanism to use spanwflag const name instead of hex or binary (more readable code)

victorbstan committed:
formatting

victorbstan committed:
formatting

victorbstan committed:
handle target exit

victorbstan committed:
formatting

victorbstan committed:
read props method, update touch helper method name

victorbstan committed:
button sounds and default refactored

victorbstan committed:
secret door sounds, animations, and props refactoring, bugfixes and cleanup

victorbstan committed:
door sounds, animations and bugfixes

victorbstan committed:
plat sound and animation , bug fixes

victorbstan committed:
train sound updates

victorbstan commented:
* Sounds for doors, secrets, plats, buttons, should respect worldtype and sounds settings correctly
* cleaned up properties fetching across the func scripts
* doors with keys have proper worldtype missing key sound
* added most if not all default properties as constants in funcs

As far as I know this is mostly feature complete for plats, doors, secret doors and buttons. 

TODO (in the future :))
* apply damage to player if blocking
* trigger when shot/damaged for some buttons/doors/secrets that take damage
* extend play of looping sound according to animation duration

victorbstan commented on quake1_example/class/_func.gd:
```diff
@@ -1,14 +1,61 @@
+	return result
+
+
+func _prop(name : StringName, def) :
```
This is used for fetching props, it also casts to default type

victorbstan committed:
cleanup

victorbstan commented on quake1_example/class/_func.gd:
```diff
@@ -1,14 +1,61 @@
 extends AnimatableBody3D
 class_name QmapbspQuakeFunctionBrush
 
+# Spawnflags:
```
Opting for constants to define spawnflags and default properties

victorbstan commented on quake1_example/class/_trigger.gd:
```diff
@@ -1,50 +1,108 @@
 
-signal emit_message_state(m : String, show : bool)
+
+func _flag(spawnflag : int) -> bool :
```
Gets the spawnflags by name

victorbstan commented on quake1_example/class/_trigger.gd:
```diff
@@ -1,50 +1,108 @@
 func _ready() :
 	v = get_meta(&'viewer')
-	monitorable = false
+	set_deferred(&'monitorable', false)
```
Saw this as suggestion by godot editor, as it caused errors sometimes when terminating app, did not see adverse effects .

victorbstan commented on quake1_example/class/_trigger.gd:
```diff
@@ -1,50 +1,108 @@
 		_trigger_now(b)
 	
+
+func _trigger_exit(b : Node3D) :
```
Trigger exist and enter are used to keep track of player for doors and plats, to check if player is still inside door trigger or plat trigger, and thus prevent closing the door or plat until player leaves trigger area

victorbstan committed:
cleanup

victorbstan commented on quake1_example/class/func_train.gd:
```diff
@@ -68,6 +53,29 @@ func _start(c : Curve3D) :
 func _f() :
 	player_end = true
 	_play_snd('plats/train2.wav')
+
```
This is mostly untouched, just formatting, moved down sound function to be consistent with other func scripts

victorbstan commented on quake1_example/wim.gd:
```diff
@@ -161,7 +161,7 @@ func _new_entity_node(classname : StringName) -> Node :
 	if node.has_signal(&'emit_message_state') :
 		node.connect(&'emit_message_state',
-			viewer._emit_message_state.bind(node)
+			viewer.emit_message_state.bind(node)
```
just removed `_` to be consistent with `emit_message_once`

gongpha commented:
Are these doors supposed to be closed after leave them ?

https://github.com/gongpha/gdQmapbsp/assets/13400398/726e3ef8-8019-4728-a591-464568e46e7c

victorbstan commented:
> Are these doors supposed to be closed after leave them ?
> Godot_v4.0.3-stable_win64.2023-07-16.01-54-51-425.mp4

What map is this?

victorbstan committed:
fixed issue with "start open" doors and with -1 doors

victorbstan commented:
> Are these doors supposed to be closed after leave them ?
> Godot_v4.0.3-stable_win64.2023-07-16.01-54-51-425.mp4

Last commit should fix this, and a couple of other door related bugs

victorbstan commented on quake1_example/class/func_door.gd:
```diff
@@ -232,17 +234,16 @@ func _trigger(b : Node3D) :
+	if _requires_key() : return
 	
-	if not _flag(TOGGLE) :
+	if not _flag(TOGGLE) and wait > 0 :
```
Don't close if -1 and exiting a trigger

victorbstan commented on quake1_example/class/func_door.gd:
```diff
@@ -265,6 +266,7 @@ func _move_close() :
 	if tween : return
 	if not open : return
 	if _is_blocked() : return
+	if not _flag(START_OPEN) and wait == -1 : return # stay open
```
if starts open, and wait is -1 it should be able to close (and stay closed)

victorbstan commented on quake1_example/class/func_door.gd:
```diff
@@ -224,6 +224,8 @@ func _is_blocked() -> bool :
 		
 		
 func _trigger(b : Node3D) :
+	if tween : return
```
Fixes infinite loop with linked doors in special cases

victorbstan committed:
new water texture name

victorbstan committed:
fix issue with buttons that are non axially aligned (bug in master)

victorbstan commented:
Found a bug that exists in `master` as well regarding non axially aligned buttons (fixed by : 
<img width="1342" alt="Screenshot 2023-07-15 at 5 23 42 PM" src="https://github.com/gongpha/gdQmapbsp/assets/315297/9c524452-aebd-4f1a-9acd-92fed275a7db">

fixed by
[694bd2c](https://github.com/gongpha/gdQmapbsp/pull/13/commits/694bd2c0237d65231c45c3fb9b66d37f3b03b541)

victorbstan committed:
simplified add calc

victorbstan committed:
fixed button add calc relative to button size

victorbstan commented:
https://github.com/gongpha/gdQmapbsp/assets/315297/8a9a56e1-2372-410e-ae75-c612a488c0f7

fixed button add calc

victorbstan committed:
fix issue with start open linked doors

victorbstan committed:
update to trigger once

victorbstan committed:
update to trigger counter

victorbstan committed:
ability to pass custom message string

victorbstan committed:
moved open check after tween check

victorbstan committed:
no message for skill setter

victorbstan committed:
fix regression and implement no touch flag for trigger once

victorbstan commented:
Latest updates: 

* various bug fixes, regression fixes
* trigger counter now shows message and triggers next target (see E1M4)
* trigger once implements NO TOUCH flag

victorbstan committed:
clenup

gongpha commented:
`trigger_onlyregistered` shows a message for unregistered users while I'm in registered mode.

![image](https://github.com/gongpha/gdQmapbsp/assets/13400398/29b0f2da-0376-4ebf-8257-013535b46757)

victorbstan committed:
regression fix for registered message

victorbstan commented:
Fixed registered message regression problem, caused by wanting to show trigger once messages

gongpha commented:
I saw you fixed some `func_train` issues. Notice 2 magic gate cubes in E1M2 are now moving synchronously to the timer.
I hope you can fix these remaining issues :

---

The moving speed is slightly slow. Notice the platform moves through the wood platforms :

https://github.com/gongpha/gdQmapbsp/assets/13400398/6732e255-fbc2-4183-a5b5-aa05de1fbd38

They're ping-pong looped on some `func_train` that has no looping :

https://github.com/gongpha/gdQmapbsp/assets/13400398/76923a6e-b09c-42e8-bad0-1fa4e48d583b

victorbstan commented:
I'll take a look at fixing those. There's probably other bugs, and I plan on fixing and implementing all missing entity features, at least to a state where it's playable, if not absolutely the same.

gongpha commented on quake1_example/class/_func.gd:
```diff
@@ -1,14 +1,61 @@
+	return result
+
+
+func _prop(name : StringName, def) :
+	var result = props.get(name, def)
+	if typeof(result) == typeof(def) : return result
+	match typeof(def) :
+		TYPE_NIL:
+			result = def
+		TYPE_BOOL:
+			result = result.to_bool()
+		TYPE_INT:
+			result = result.to_int()
+		TYPE_FLOAT:
+			result = result.to_float()
+		TYPE_STRING:
+			result = str(result)
+		_ :
+			result = def
+	return result
```
I've seen this one in `_trigger`. Would it be better if we move this function to be a static function in a util class which any class can be accessed ?

gongpha replied:
I recommended using regular strings as a dictionary key. Since the StringNames and strings are considered the same. So the GDScript wouldn't take time to construct a new StringName for a key. (Constructing a new StringName is slightly slower than a regular string. except the case if the StringName is a literal form like `&"abc"`)

victorbstan replied:
Aren't StringNames better for repetitive keys, so they point to the same memory space instead of alocating new memory for all key names? and for the second part `... case if the StringName is a literal form like &"abc")` -- isn't that the case?

gongpha replied:
Godot dictionary may not work as you think. Try running this snippet.
```gdscript
var dict := {}
dict[&'a'] = 1
print(dict.find_key(1) is StringName)
```
This will print `false`.

Godot will internally convert the StringName keys to strings. I think in practice, StringName and string have the same content and their hash should be the same too. Sometimes I argue against this behavior too. It might confuse end users.

But originally, the StringNames were exposed to GDScript in 4.0 for exchanging some engine properties that have unique names e.g. node names, method names, and signal names.

victorbstan replied:
Seems it's not like ruby's hash/dictionary then, missed opportunity IMHO -- I'll change dict keys to strings then :)

gongpha commented on quake1_example/class/func_door.gd:
```diff
@@ -4,77 +4,115 @@ class_name QmapbspQuakeFunctionDoor
+# 0 : "Medieval"
+# 1 : "Metal (runic)"
+# 2 : "Base"
+const gold_keys_message = [
```
If the keys are required to open doors, that means we have to implement key entities. Isn't it ? ^^"

victorbstan replied:
Of course :) this just makes the doors behave as if that was the case. We handle keys and such in an inventory for the player, but that's not for this PR IMHO.

victorbstan replied:
We will implement keys, and pickups, and everything else at some point :)

victorbstan committed:
switch _prop name, prop.get, prop.has; to String from StringName

victorbstan committed:
added sounds and plat default constants, WIP on proper positioning along path

victorbstan committed:
Merge branch 'master' into doors-platforms-triggers

victorbstan committed:
bug fixes to audio and collision detection after merge updates

victorbstan committed:
Merge branch 'master' into doors-platforms-triggers

victorbstan committed:
changed corner path list from curve to array of path objects, fixed train timming and added wait for each point

victorbstan commented:
Platforms now should move to correct time/duration and wait according to `wait` settings for each path corner.

gongpha commented:
Strangely sounds playback when touching doors.

https://github.com/gongpha/gdQmapbsp/assets/13400398/186d1698-0346-429b-9674-503922c9c1cd

victorbstan committed:
audio fixes to allow custom override for interrupting current audio playing
